src/main/scala/arcana/empress/Lexer.scala

```diff
@@ -0,0 +1,90 @@
+package arcana.empress
+
+import arcana.empress._
+
+class Lexer(code: String, options: LexerOptions) {
+  private type LocatedToken = (Token, Option[Location])
+  lazy val tokens: List[LocatedToken] = parse()
+  
+  private def parse(): List[LocatedToken] =
+    code.split("\n").iterator.zipWithIndex.map(
+      {case (line, i) => parseLine(List(), 0)(line, i + 1)}).toList.flatten
+
+  import Lexer.{keywords, punctuations}
+  private def parseLine(parsed: List[LocatedToken], column: Int = 0)(implicit str: String, line: Int): List[LocatedToken] =
+    if (column == str.length()) parsed
+    else {
+      if (str(column) == ';') parsed // comments
+      else if (punctuations.contains(str(column))) {
+        if (str(column).isSpaceChar) parseLine(parsed, column + 1) // skip space chars
+        else parseLine(parsed :+ parsePunctuation(str(column), Position(line, column)), column + 1)
+      }
+      else if (str(column) == '"') parseStringLiteral(str, Position(line, column)) match {
+        case (locatedToken, next) => parseLine(parsed :+ locatedToken, next)
+      }
+      else if (str(column).isDigit) parseNumericLiteral(str, Position(line, column)) match {
+        case (locatedToken, next) => parseLine(parsed :+ locatedToken, next)
+      }
+      else parseIdentifierAndKeyword(str, Position(line, column)) match {
+        case (locatedToken, next) => parseLine(parsed :+ locatedToken, next)
+      }
```

Comment:
why use three match here

Reply:
Deconstruct the pair result without creating a temp variable.

Reply:
but you always return same pair result, way just follow with parseLine without match

Reply:
What?